### PR TITLE
GH-73: As a developer I need "find references" to work also in corner cases

### PR DESCRIPTION
**Pull request by qtran-n4**
## Task
Link: https://github.com/eclipse/n4js/issues/73

### CODE REVIEW

**Comment by qtran-n4** [comment]
@mor-n4 : Please start to review, especially the part related to cross refs and imported names. Thanks!

File: plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/utils/TypeHelper.xtend [hunk 1]
```diff
@@ -194,4 +199,42 @@ public class TypeHelper {
 		return -1;
 	}
 
+	/**
+	 * Check if an EObject is a composed member. GH-73: TODO: duplicate with {@code ComposedMemberScope#isComposedMember}.
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/utils/TypeHelper.xtend]: remove one

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/ASTMetaInfoCache.java [hunk 2]
```diff
@@ -109,9 +109,12 @@ public boolean hasBrokenAST() {
 			if (isCanceled()) {
 				return new Result<>(TypeRefsFactory.eINSTANCE.createUnknownTypeRef());
 			} else {
-				throw UtilN4.reportError(new IllegalStateException(
-						"cache miss: no actual type in cache for AST node: " + astNode
-								+ " in resource: " + resource.getURI()));
+				// TODO GH-73: Temporary hack. Reactivate the throw statement and solve cache miss problem, i.e. GH-95
+				Result<TypeRef> tr = new Result<>((TypeRef) astNode.eContainer());
+				return tr;
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/ASTMetaInfoCache.java]: See this bug: https://github.com/eclipse/n4js/issues/95 :)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/resource/N4JSCrossReferenceComputer.java [hunk 3]
```diff
@@ -58,25 +62,43 @@
...
 	 *            the logic that collects the passed EObject found in a cross reference
 	 */
-	void computeCrossRefs(Resource resource, IAcceptor<EObject> acceptor) {
-		TreeIterator<EObject> allContentsIter = resource.getAllContents();
+	public void computeCrossRefs(Resource resource, IAcceptor<ImmutablePair<EObject, EObject>> acceptor) {
+		TreeIterator<EObject> allContentsIter;
+		if (resource instanceof N4JSResource) {
+			allContentsIter = ((N4JSResource) resource).getAllContentIgnoreCachedElements();
+		} else {
+			allContentsIter = resource.getAllContents();
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/resource/N4JSCrossReferenceComputer.java]: Hmm, I think you are right. I don't see any reason for traversing the `TModule` either...

File: plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/findReferences/ConcreteSyntaxAwareReferenceFinder.java [hunk 4]
```diff
@@ -105,4 +110,99 @@ protected EObject toValidInstanceOrNull(Resource resource, Predicate<URI> target
...
+
+	/**
+	 * This method overrides Xtext's standard behavior because we need special handling for composed members.
+	 */
+	@Override
+	protected void findLocalReferencesFromElement(
+			Predicate<URI> targetURIs,
+			EObject sourceCandidate,
+			Resource localResource,
+			Acceptor acceptor) {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/findReferences/ConcreteSyntaxAwareReferenceFinder.java]: at this point: make sure we do not traverse the TModule if in a N4JSResource
e.g.
```
		if (sourceCandidate instanceof TModule) {
			return;
		}
```
(but check for better location)

File: plugins/org.eclipse.n4js.ts.ui/src/org/eclipse/n4js/ts/ui/search/LabellingReferenceFinder.java [hunk 5]
```diff
@@ -37,12 +39,16 @@ protected Acceptor toAcceptor(IAcceptor<IReferenceDescription> acceptor) {
...
 			public void accept(EObject source, URI sourceURI, EReference eReference, int index, EObject targetOrProxy,
 					URI targetURI) {
-				String name = labelProvider.getText(source);
-				LabelledReferenceDescription description = new LabelledReferenceDescription(sourceURI, targetURI,
-						eReference, index, name);
-				accept(description);
+				ICompositeNode srcNode = NodeModelUtils.getNode(source);
+				if (srcNode != null) {
+					// Only accept candidates with a corresponding node model.
+					// GH-73: TODO improve this
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ts.ui/src/org/eclipse/n4js/ts/ui/search/LabellingReferenceFinder.java]: probably no longer required, after making sure that TModule and "Ignored cached elements" are not traversed in ConcreteSyntaxAwareReferenceFinder

**Comment by mor-n4** [comment]
example from our discussion:
```
// B.n4js
import {x} from "A";
x.foo(null);
```
and
```
// A.n4js
export public interface I {
	public foo(p: I);
}
export public interface J {
	public foo(p: I);
}
export var x: I|J;
```
Now: search for references of `I`.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/ASTMetaInfoCache.java (hunk 2 above)
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/ASTMetaInfoCache.java]: but `(TypeRef) astNode.eContainer()` will throw class cast exceptions, most of the time, right?
Does #95 block #73? Then we need to fix it.

Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/ASTMetaInfoCache.java]: Yes, https://github.com/eclipse/n4js/issues/95 blocks https://github.com/eclipse/n4js/issues/73 . We will definitely need to fix https://github.com/eclipse/n4js/issues/95 before being able to merge https://github.com/eclipse/n4js/issues/73

File: plugins/org.eclipse.n4js.ts.ui/src/org/eclipse/n4js/ts/ui/search/LabellingReferenceFinder.java (hunk 5 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ts.ui/src/org/eclipse/n4js/ts/ui/search/LabellingReferenceFinder.java]: done

File: plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/findReferences/ConcreteSyntaxAwareReferenceFinder.java (hunk 4 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/findReferences/ConcreteSyntaxAwareReferenceFinder.java]: done

**Comment by qtran-n4** [comment]
@mor-n4 This PR is ready to merge. Please review the changes again. Thanks!

**Comment by qtran-n4** [comment]
@mor-n4 The PR is finally almost ready :-) Please review already. I am still cleaning up the last remaining minor TODOs. 

Compared to the first review, this PR incorporates the following changes, among others

- Refactor `N4JSCrossReferenceComputer` to not not pass the AST node to the acceptor. Also the handling case for EReference TYPEREF is removed since it never happens.
- Improve `ComposedMemberScope#getOrCreateComposedMemberCache` to search for existing cache in the current module of equivalent composed member ref type
- Refactor `N4JSExpressionValidator` so that no member scope is created during the validation of the constructor visibility eliminating one case of calling `MemberScopingHelper#createMemberScopeAllowingNonContainedMembers`. Instead this is done during scope (See `N4JSScopeProvider#scope_IdentifierRef_id`
- Improve the performance of find references by optimizing the check if it is worth it to look into a resource in `ConcreteSyntaxAwareReferenceFinder#findReferencesInDescription` 
- Refactor logics for Cmd + click of composed members to use Xtext's infrastructure instead of having to implement a dialog ourselves (See `ComposedMemberAwareHyperlinkHelper`)

File: plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/utils/TypeHelper.xtend [hunk 6]
```diff
@@ -194,4 +200,46 @@ public class TypeHelper {
...
 
+
+	/**
+	 * Extracts the type(s) referenced by a type ref.
+	 * @param typeRef
+	 *  		the type ref, from which the type should be extracted.
+	 * @return
+	 * 			the list of extracted types. There can be multiple returned types due to composed type refs.
+	 */
+	public def List<Type> extractType(TypeArgument typeArg) {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/utils/TypeHelper.xtend]: 1. please double-check that we do not already have something like this, somewhere.
2. rename to `extractTypes` (or maybe `collectTypes`).
3. this does not seem to require injection, so make static and move to `TypeUtils`.

File: plugins/org.eclipse.n4js.ui/META-INF/MANIFEST.MF [hunk 7]
```diff
@@ -46,7 +46,8 @@ Require-Bundle: org.eclipse.n4js;visibility:=reexport,
  com.fasterxml.jackson.core.jackson-core,
  com.fasterxml.jackson.core.jackson-databind,
  org.eclipse.n4js.utils.ui,
- com.ibm.icu
+ com.ibm.icu,
+ org.eclipse.search
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ui/META-INF/MANIFEST.MF]: why these new dependencies?
do we have CQs for them?

File: plugins/org.eclipse.n4js.ts.model/model/Types.xcore [hunk 8]
```diff
@@ -161,6 +161,8 @@ class TModule extends SyntaxRelatedTElement, TAnnotableElement {
  */
 class ComposedMemberCache {
 	contains transient TMember[] cachedComposedMembers
+	// The compose type ref of the member being cached
+	contains transient TypeRef composedTypeRef
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ts.model/model/Types.xcore]: why containment?
if we use containment here, it is important to use TypeUtils.copyIfContained() before setting the value of `composedTypeRef`.

**Comment by mor-n4** [comment]
what about the copyright of the newly added icon?

File: plugins/org.eclipse.n4js.xpect/src/org/eclipse/n4js/xpect/common/N4JSOffsetAdapter.java [hunk 9]
```diff
@@ -41,9 +41,9 @@
 	 * {@link EObjectCoveringRegionProvider}. The AST related node covers the complete region defined by the offset of
 	 * the specified location and the length of the specified location. That is, for a given test class with a parameter
 	 * 'of' as OFFSET (via {@code @ParameterParser(syntax = "('of' arg1=OFFSET)?")}), the following two tests <br/>
-	 * {@code // XPECT ... of 'a as B'} <br/>
+	 * {@code // XpECT ... of 'a as B'} <br/>
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.xpect/src/org/eclipse/n4js/xpect/common/N4JSOffsetAdapter.java]: /sadface

**Comment by qtran-n4** [comment]
The icons are all Eclipse icons :-)

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend [hunk 10]
```diff
@@ -280,6 +287,13 @@ class N4JSScopeProvider extends AbstractScopeProvider implements IDelegatingScop
 	 */
 	private def IScope scope_IdentifierRef_id(IdentifierRef identifierRef, EReference ref) {
 		val VariableEnvironmentElement vee = identifierRef.ancestor(VariableEnvironmentElement);
+		val scope = getLexicalEnvironmentScope(vee, identifierRef, ref);
+		// Handle constructor visibility
+		if (identifierRef.eContainer instanceof NewExpression) {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend]: at first sight, this special case seems a bit odd to me, at this place. But maybe I just don't get it. Let's discuss later...

File: plugins/org.eclipse.n4js/META-INF/MANIFEST.MF [hunk 11]
```diff
@@ -35,6 +35,7 @@ Require-Bundle: org.eclipse.xtext;visibility:=reexport,
  org.eclipse.n4js.external.libraries,
  org.eclipse.core.jobs,
  org.eclipse.core.resources
+Import-Package: org.apache.commons.lang3.tuple
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/META-INF/MANIFEST.MF]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/resource/N4JSResourceDescription.java [hunk 12]
```diff
@@ -57,7 +60,9 @@
 
 	private final IDefaultResourceDescriptionStrategy strategy;
 
-	private Iterable<QualifiedName> lazyImportedNames;
+	private SortedSet<QualifiedName> lazyImportedNames;
+
+	// private final N4JSTypeSystem ts;
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/resource/N4JSResourceDescription.java]: done

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/misc/DestructureHelper.xtend [hunk 13]
```diff
@@ -151,6 +150,8 @@ class DestructureHelper {
 		}
 		else {
 			// non-positional
+			// TODO: Consider implementing scoping for binding pattern's properties or use other validation instead of
+			// creating scope here
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/misc/DestructureHelper.xtend]: Removed TODO

File: plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/utils/TypeHelper.xtend (hunk 6 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ts/src/org/eclipse/n4js/ts/utils/TypeHelper.xtend]: this method is not used anymore. Removed it.

File: plugins/org.eclipse.n4js.ts.model/model/Types.xcore (hunk 8 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ts.model/model/Types.xcore]: Change to `refers transient TypeRef composedTypeRef`

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend (hunk 10 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend]: This is the place where scoping is done for IdentifierRef in a new expression

File: plugins/org.eclipse.n4js.n4jsx.ui/src/org/eclipse/n4js/n4jsx/ui/search/N4JSXReferenceQueryExecutor.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+				// Do not show JSX element's closing tag
+				return false;
+			} else {
+				return true;
+			}
+		} else {
+			return super.isRelevantToUser(referenceDescription);
+		}
+	}
+
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.n4jsx.ui/src/org/eclipse/n4js/n4jsx/ui/search/N4JSXReferenceQueryExecutor.java]: done

File: plugins/org.eclipse.n4js.ui/META-INF/MANIFEST.MF (hunk 7 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ui/META-INF/MANIFEST.MF]: com.ibm.icu dependency has been there for a while.

File: plugins/org.eclipse.n4js.xpect/src/org/eclipse/n4js/xpect/common/N4JSOffsetAdapter.java (hunk 9 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.xpect/src/org/eclipse/n4js/xpect/common/N4JSOffsetAdapter.java]: -> Moritz: /sadface

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend (hunk 10 above)
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend]: as discussed:
leave as is, for now; created separate ticket #164

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java [hunk 15]
```diff
@@ -246,8 +247,16 @@ private ComposedMemberCache getOrCreateComposedMemberCache() {
...
+				// Search in the module for composed member cache containing equivalent composed type ref
+				for (ComposedMemberCache existingCache : module.getComposedMemberCaches()) {
+					if (TypeCompareUtils.isEqual(existingCache.getComposedTypeRef(), composedTypeRef)) {
+						return existingCache;
+					}
+				}
+
 				final ComposedMemberCache cacheNew = TypesFactory.eINSTANCE.createComposedMemberCache();
 				EcoreUtilN4.doWithDeliver(false, () -> {
+					cacheNew.setComposedTypeRef(composedTypeRef);
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java]: add comment at end of this line: `// order important!!`

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java [hunk 16]
```diff
@@ -246,8 +250,17 @@ private ComposedMemberCache getOrCreateComposedMemberCache() {
 			final Resource res = contextCasted.eResource();
 			final TModule module = res instanceof N4JSResource ? ((N4JSResource) res).getModule() : null;
 			if (module != null) {
+				// Search in the module for composed member cache containing equivalent composed type ref
+				for (ComposedMemberCache existingCache : module.getComposedMemberCaches()) {
+					if (typeCompareHelper.getTypeRefComparator().compare(existingCache.getComposedTypeRef(),
+							composedTypeRef) == 0) {
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java]: Done

Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java]: Done

**Comment by qtran-n4** [comment]
@mor-n4 : I have incorporated your comments in the new commits. Please have a look again. Thanks!

File: plugins/org.eclipse.n4js.ts.model/model/Types.xcore (hunk 8 above)
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.ts.model/model/Types.xcore]: reversed back to `contains transient`

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java [hunk 17]
```diff
@@ -239,15 +240,24 @@ private ComposedMemberCache getOrCreateComposedMemberCache() {
 			final MemberAccess contextCasted = //
 					(MemberAccess) request.context; // cast is valid, see MemberScopeRequest#provideContainedMembers
 			final ComposedMemberCache cache = contextCasted.getComposedMemberCache();
-			if (cache != null) {
+			if (cache != null && !TypeCompareUtils.isEqual(cache.getComposedTypeRef(), this.composedTypeRef)) {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java]: wrong logic; the `!` must be removed

Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/members/ComposedMemberScope.java]: done

**Comment by qtran-n4** [comment]
Thanks for for review. The builds are green. Merging now.

**Pull request merged**